bakery/authorizer.go

```diff
@@ -0,0 +1,134 @@
+	//
+	// On success, each element of allowed holds whether the respective
+	// element of ops has been allowed, and caveats holds any additional
+	// third party caveats that apply.
+	Authorize(ctxt context.Context, id Identity, ops []Op) (allowed []bool, caveats []checkers.Caveat, err error)
+}
+
+var (
+	// OpenAuthorizer is an Authorizer implementation that will authorize all operations without question.
+	OpenAuthorizer openAuthorizer
+
+	// ClosedAuthorizer is an Authorizer implementation that will return ErrPermissionDenied
+	// on all authorization requests.
+	ClosedAuthorizer closedAuthorizer
+)
+
+type openAuthorizer struct{}
+
+func (openAuthorizer) Authorize(ctxt context.Context, id Identity, ops []Op) (allowed []bool, caveats []checkers.Caveat, err error) {
+	allowed = make([]bool, len(ops))
+	for i := range allowed {
+		allowed[i] = true
+	}
+	return allowed, nil, nil
+}
+
+type closedAuthorizer struct{}
+
+func (closedAuthorizer) Authorize(ctxt context.Context, id Identity, ops []Op) (allowed []bool, caveats []checkers.Caveat, err error) {
+	return make([]bool, len(ops)), nil, nil
+}
+
+type authInfoKey struct{}
+
+func ContextWithAuthInfo(ctxt context.Context, authInfo *AuthInfo) context.Context {
+	return context.WithValue(ctxt, authInfoKey{}, authInfo)
+}
+
+func AuthInfoFromContext(ctxt context.Context) *AuthInfo {
+	authInfo, _ := ctxt.Value(authInfoKey{}).(*AuthInfo)
+	return authInfo
+}
+
+// ACLAuthorizer is an Authorizer implementation that will check ACL membership
+// of users. It uses GetACLs to find out the ACLs that apply to the requested
+// operations and will authorize an operation if an ACL contains the
+// group "everyone" or if the context contains an AuthInfo (see
+// ContextWithAuthInfo) that holds an Identity that implements
+// ACLIdentity and its Allow method returns true for the ACL.
+type ACLAuthorizer struct {
+	// If AllowPublic is true and an ACL contains "everyone",
+	// then authorization will be granted even if there is
+	// no logged in user.
+	AllowPublic bool
+
+	// GetACLs returns the ACL that applies to each of the given
+	// operations. It should return a slice with each element
+	// holding the ACL for the corresponding operation in the
+	// argument slice.
+	//
+	// If an entity cannot be found or the action is not recognised,
+	// GetACLs should return an empty ACL entry for that operation.
+	GetACLs func(ctxt context.Context, ops []Op) ([][]string, error)
+}
+
+// ACLIdentity may be implemented by Identity implementions
+// to report group membership information.
+// See ACLAuthorizer for details.
+type ACLIdentity interface {
+	Identity
+
+	// Allow reports whether the user should be allowed to access
+	// any of the users or groups in the given ACL slice.
+	Allow(ctxt context.Context, acl []string) (bool, error)
+}
+
+func (a ACLAuthorizer) Authorize(ctxt context.Context, id Identity, ops []Op) (allowed []bool, caveats []checkers.Caveat, err error) {
+	if ops == nil {
```

Comment:
Probably, but we're being defensive. From the point of view of the authorizer, we're asking whether the user can do nothing, and the answer is yes.

But I actually didn't intend to include this type in this PR. I'll remove it and add it later as it hasn't got decent tests yet anyway.